### PR TITLE
techdebt: [CDS-38177]: Referncing the userGroups

### PR DESCRIPTION
**Pull request by Y0709**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/34105)

### CODE REVIEW

**Comment by Y0709** [comment]
trigger messagecheck

File: 400-rest/src/main/java/software/wings/service/impl/compliance/GovernanceConfigServiceImpl.java [hunk 1]
```diff
@@ -237,6 +240,45 @@ public GovernanceConfig upsert(@AccountId String accountId, @Nonnull GovernanceC
...
+    return referencedUserGroups;
+  }
+
+  private void updateFreezeWindowReferenceInUserGroup(
+      Set<String> previousUserGroups, Set<String> currentUserGroups, String accountId, String appId, String EntityId) {
+    Set<String> parentsToRemove = Sets.difference(previousUserGroups, currentUserGroups);
+    Set<String> parentsToAdd = Sets.difference(currentUserGroups, previousUserGroups);
+
+    for (String id : parentsToRemove) {
+      userGroupService.removeParentsReference(id, accountId, appId, EntityId, "GOVERNANCE_CONFIG");
```
Comment (abhinav-harness) [inline 400-rest/src/main/java/software/wings/service/impl/compliance/GovernanceConfigServiceImpl.java:275]: small E in entity id

File: 400-rest/src/main/java/software/wings/service/impl/compliance/GovernanceConfigServiceImpl.java [hunk 2]
```diff
@@ -237,6 +239,45 @@ public GovernanceConfig upsert(@AccountId String accountId, @Nonnull GovernanceC
     }
   }
 
+  private void updateUserGroupReference(
+      GovernanceConfig updatedSetting, GovernanceConfig oldSetting, String accountId) {
+    List<TimeRangeBasedFreezeConfig> oldTimeRangeBasedFreezeConfigs = oldSetting.getTimeRangeBasedFreezeConfigs();
+    List<TimeRangeBasedFreezeConfig> newTimeRangeBasedFreezeConfigs = updatedSetting.getTimeRangeBasedFreezeConfigs();
```
Comment (abhinav-harness) [inline 400-rest/src/main/java/software/wings/service/impl/compliance/GovernanceConfigServiceImpl.java:245]: add to execption msg in usergroupserviceimpl about ref in deployment freeze

**Comment by Y0709** [comment]
trigger codebasehash

**Review by hingerabhinav — APPROVED** [review]
> Reviewed 31 of 31 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34105)* status: :shipit: complete! ✓

**Comment by Y0709** [comment]
trigger ft1 ,ft2

**Review by hingerabhinav — DISMISSED** [review]
> Reviewed 71 of 71 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34105)* status: :shipit: complete! ✓

___
*[`390-db-migration/src/main/java/io/harness/migrations/all/AddDeploymentFreezeReferenceToUserGroups.java` line 54 at r5](https://reviewable.io/reviews/harness/harness-core/34105#-N4bSfIHBKi1P-cY5QOv:-N4bSfIHBKi1P-cY5QOw:b-gwdnx2) ([raw file](https://github.com/harness/harness-core/blob/afa30573d43097ed92b762e7c57ae63dbd52ef35/390-db-migration/src/main/java/io/harness/migrations/all/AddDeploymentFreezeReferenceToUserGroups.java#L54)):*
> ```Java
>           governanceConfigService.getReferencedUserGroupIds(governanceConfig.getTimeRangeBasedFreezeConfigs());
>       for (String id : userGroupIds) {
>         userGroupService.addParentsReference(
> ```

We're doing 1 API call for each user group that is used in a deployment freeze. 
Please check the number of user groups in QA/Prod per account.

**Review by abhinav-harness — CHANGES_REQUESTED** [review]
> Reviewed 4 of 31 files at r4, 46 of 71 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34105)* status: :shipit: complete! ✓

___
*[`400-rest/src/main/java/software/wings/service/impl/compliance/GovernanceConfigServiceImpl.java` line 254 at r5](https://reviewable.io/reviews/harness/harness-core/34105#-N4fNuxc6_BIW9xOzvYF:-N4fNuxc6_BIW9xOzvYG:b-n3ltsk) ([raw file](https://github.com/harness/harness-core/blob/afa30573d43097ed92b762e7c57ae63dbd52ef35/400-rest/src/main/java/software/wings/service/impl/compliance/GovernanceConfigServiceImpl.java#L254)):*
> ```Java
>     } catch (Exception e) {
>       log.error(
>           "error while fetching the timeRangeBasedFreezeConfig in account with id {} in Governance Config with id {}",
> ```

add user group info too

**Review by abhinav-harness — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34105)* status: :shipit: complete! ✓

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34105)* status: :shipit: complete! ✓

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 22 of 71 files at r5.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34105)* status: :shipit: complete! ✓

**Review by poojaSinghal — CHANGES_REQUESTED** [review]
> Reviewed 3 of 31 files at r4, 47 of 71 files at r5, 1 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34105)* status: :shipit: complete! ✓

___
*[`390-db-migration/src/main/java/io/harness/migrations/all/AddDeploymentFreezeReferenceToUserGroups.java` line 47 at r6](https://reviewable.io/reviews/harness/harness-core/34105#-N4gI9AEA395nYM2mKPP:-N4gI9AF5GZPyQSRMhFY:b-7cfvrg) ([raw file](https://github.com/harness/harness-core/blob/4235d56f8711d24b8916089b22fad2b3f087dc28/390-db-migration/src/main/java/io/harness/migrations/all/AddDeploymentFreezeReferenceToUserGroups.java#L47)):*
> ```Java
> 
>   private void migrateDeploymentFreezeWindowsForAccount(Account account) {
>     try {
> ```

add log line
___
*[`390-db-migration/src/main/java/io/harness/migrations/all/AddDeploymentFreezeReferenceToUserGroups.java` line 52 at r6](https://reviewable.io/reviews/harness/harness-core/34105#-N4gIAPD8k-NaPKZNyKO:-N4gIAPD8k-NaPKZNyKP:b555i5p) ([raw file](https://github.com/harness/harness-core/blob/4235d56f8711d24b8916089b22fad2b3f087dc28/390-db-migration/src/main/java/io/harness/migrations/all/AddDeploymentFreezeReferenceToUserGroups.java#L52)):*
> ```Java
>                                               .get();
>       Set<String> userGroupIds =
>           governanceConfigService.getReferencedUserGroupIds(governanceConfig.getTimeRangeBasedFreezeConfigs());
> ```

add null check

**Pull request merged**